### ISSUE TITLE
output always nan

### ISSUE THREAD

**Issue by zekageri** [Dlloydev/sTune#28]
First of all, great library and thank you for the effort.

I'm using an ESP32 Arduino.

I have an implementation where a user can create any number of `thermSystems` which is represented as a class for me.
In every `thermSystem` there is it's own `set` and `measured` temperature coming from different sensors.
For every `thermSystem` there is only one `boiler` with it's own input which controlls if it has to start to heat or not.
For me, the esp can not directly set a pin to high or low, so i modified your library a little bit in the following ways:

**sTune.h**
```cpp
float softDigit(boolean &heatControllBit, float input, float output, float setpoint, uint32_t windowSize, uint8_t debounce);
```
**sTune.cpp**

```cpp
float sTune::softDigit(boolean &heatControllBit, float input, float output, float setpoint, uint32_t windowSize, uint8_t debounce) {

  // software PWM timer
  uint32_t msNow = millis();
  static uint32_t  windowStartTime, nextSwitchTime;
  if (msNow - windowStartTime >= windowSize) {
    windowStartTime = msNow;
  }
  // SSR optimum AC half-cycle controller
  static float optimumOutput;
  static bool reachedSetpoint;

  if (input > setpoint) reachedSetpoint = true;
  if (reachedSetpoint && !debounce && setpoint > 0 && input > setpoint) optimumOutput = output - 8;
  else if (reachedSetpoint && !debounce && setpoint > 0 && input < setpoint) optimumOutput = output + 8;
  else  optimumOutput = output;
  if (optimumOutput < 0) optimumOutput = 0;

  // PWM relay output
  static bool relayStatus;
  if (!relayStatus && optimumOutput > (msNow - windowStartTime)) {
    if (msNow > nextSwitchTime) {
      nextSwitchTime = msNow + debounce;
      relayStatus = true;
      heatControllBit = true;
    }
  } else if (relayStatus && optimumOutput < (msNow - windowStartTime)) {
    if (msNow > nextSwitchTime) {
      nextSwitchTime = msNow + debounce;
      relayStatus = false;
      heatControllBit = false;
    }
  }
  return optimumOutput;
}
```

As you can see i have added a modification to the softpwm function to be able to set a `boolean`  variable by `value` instead of a `digitalWrite`. This is to ensure that no pin is changed in my setup and controlls only at software level.

So with that in mind, i have the following controll functions for the auto tune PID.

```cpp
/* PID DETAILS */
const uint8_t inputPin = 0;

uint32_t settleTimeSec = 10;
uint32_t testTimeSec = 1000;
const uint16_t samples = 500;
const float inputSpan = 150;
const float outputSpan = 1000;
float outputStart = 0;
float outputStep = 300;
float tempLimit = 350;

// variables
float Input, Output, Setpoint = 50, Kp, Ki, Kd;
sTune *tuner;
QuickPID *myPID;

void pidInit();
void pidLoop();
/* PID DETAILS */

/*
*     This function is called at startup of the thermSystem
*/
void thermSystem::pidInit() {
    tuner = new sTune(&Input, &Output, tuner->ZN_PID, tuner->directIP, tuner->printOFF);
    myPID = new QuickPID(&Input, &Output, &Setpoint);
    tuner->Configure(inputSpan, outputSpan, outputStart, outputStep, testTimeSec, settleTimeSec, samples);
    tuner->SetEmergencyStop(tempLimit);
}

/*
*     This function is called in every iteration of the loop()
*/
void thermSystem::pidLoop() {
    tuner->softDigit(shouldStartHeat, Input, Output, 0, outputSpan, 1);

    switch (tuner->Run()) {
        case tuner->sample:  // active once per sample during test
            Input = currentTemp;
            //Input = analogRead(inputPin) * 0.322265625 - 50.0;  // get degC (using 3.3v AREF)
            tuner->plotter(Input, Output * 0.1, Setpoint, 1, 3);
            break;

        case tuner->tunings:                       // active just once when sTune is done
            tuner->GetAutoTunings(&Kp, &Ki, &Kd);  // sketch variables updated by sTune
            myPID->SetOutputLimits(0, outputSpan);
            myPID->SetSampleTimeUs(outputSpan * 1000 - 1);
            myPID->SetMode(myPID->Control::automatic);  // the PID is turned on
            myPID->SetProportionalMode(myPID->pMode::pOnMeas);
            myPID->SetAntiWindupMode(myPID->iAwMode::iAwClamp);
            myPID->SetTunings(Kp, Ki, Kd);  // update PID with the new tunings
            break;

        case tuner->runPid: // active once per sample after tunings
            //Input = analogRead(inputPin) * 0.322265625 - 50.0;  // get degC (using 3.3v AREF)
            Input = currentTemp;
            myPID->Compute();
            tuner->plotter(Input, Output, Setpoint, 0.1f, 3);
            break;
    }
}
```

I have the following output on the serial monitor:

```cpp
Setpoint:150.00, Input:197.00, Output:nan,
Setpoint:225.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:230.00, Input:0.00, Output:nan,
Setpoint:150.00, Input:198.00, Output:nan,
Setpoint:225.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:230.00, Input:0.00, Output:nan,
Setpoint:150.00, Input:198.00, Output:nan,
Setpoint:225.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:230.00, Input:0.00, Output:nan,
Setpoint:150.00, Input:199.00, Output:nan,
Setpoint:225.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:230.00, Input:0.00, Output:nan,
Setpoint:150.00, Input:199.00, Output:nan,
Setpoint:225.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:210.00, Input:0.00, Output:nan,
Setpoint:230.00, Input:0.00, Output:nan,
Setpoint:150.00, Input:199.00, Output:nan,
Setpoint:225.00, Input:0.00, Output:nan,
```

And my heating can not turn off. 
It is always on but the current set temperature is 15.5 celsius which is represented as 155 decimal and the current temperature is 20 celsius which is 200 decimal.

current temperature value is coming in Serial modbus communication which sample rate is roughly equal to the loop iteration in my setup.

**What could be the problem?**

**EDIT** 

If i restart the system with 155 set temperature and 200 measured temperature the pid is turning on and off my boolean variable for a while, even if the set temperature never exceeds the measured.

**At the start of the system it looks like this:**

```cpp
Setpoint:225.00, Input:0.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:41.23,
Setpoint:210.00, Input:0.00, Output:40.47,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:58.07,
Setpoint:210.00, Input:0.00, Output:56.18,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:74.91,
Setpoint:210.00, Input:0.00, Output:71.89,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:91.75,
Setpoint:210.00, Input:0.00, Output:87.60,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:100.00,
Setpoint:210.00, Input:0.00, Output:100.00,
Setpoint:230.00, Input:0.00, Output:30.00,
Setpoint:150.00, Input:202.00, Output:30.00,
Setpoint:210.00, Input:0.00, Output:30.00,
Setpoint:225.00, Input:0.00, Output:100.00,
Setpoint:210.00, Input:0.00, Output:100.00,
Setpoint:230.00, Input:0.00, Output:30.00,
```

Also there is `valve open time` which i have to wait before i start the boiler. I don't know how to integrate it into the PID, since it can vari between 1 min to 5 or 10 min while the valve is opening.

**Comment by Dlloydev** [comment]
Thank you for your detailed information on your system and tests.

I've created an [example in Wokwi](https://wokwi.com/projects/355040432629843969) perhaps you could look at.
Your function seems to be linking OK to the `boolean shouldStartHeat;` variable and the Output isn't showing nan.

Still, on your system, you may need to use a global variable in the in the class ... `*_shouldStartHeat` in a way similar to how `*_input` and `*_output` is used in the header file and .cpp file.

The valve open time which varies 60 to 600 seconds I think would be the deadtime. This would be 6 to 60% of your test time  `uint32_t testTimeSec = 1000;`.(use a testTimeSec value that is an estimate of the time constant of your system). Here's a [chart](https://blog.opticontrols.com/wp-content/uploads/2011/06/td-versus-tau.png) that shows how difficult a system will be to control depending on the ratio of deadtime to tau (time constant). You might have to use PI control rather than PID to get more range of controllability.

**Comment by zekageri** [comment]
Thank you for the response.
The shouldStartHeat is indeed global in the class, but it isn't a pointer. I will try this and will take a look at the example. Thank you very much. As of the valve time. It is an input which a user can control depending on he's phisycal valve type. So I think I will make the dead time equal to this time.

Will test your suggestions and come back to you!

**Comment by zekageri** [comment]
In the example you created at wokwi, the led is just blinking and does not stay on.

Shouldnt it stay on to let the heating heat?

**Comment by zekageri** [comment]
Do you have any idea why is my input sometimes 0 when plotting to Serial? I have consistent readings

```
Setpoint:230.00, Input:0.00, Output:5.00,
Setpoint:210.00, Input:0.00, Output:5.00,
Setpoint:170.00, Input:176.00, Output:5.00,
Setpoint:225.00, Input:0.00, Output:5.00,
Setpoint:210.00, Input:0.00, Output:5.00,
Setpoint:230.00, Input:0.00, Output:5.00,
Setpoint:210.00, Input:0.00, Output:5.00,
Setpoint:170.00, Input:178.00, Output:5.00,
```

**Comment by zekageri** [comment]
Really confused about why the `Setpoint` and `Input` is changing while these should be constants??

I have enabled every serial output and here is the results. Right now the set temperature is 200, the measured is 191 and it does not set the boolean value to true at all. Sometimes it does set it but it does not turn it off ever.

```cpp
Setpoint:225.00, Input:0.00, Output:0.00,
 sec: 140.1491  out: 100.00  pv: 0.000  tan: -111.500 →
 sec: 145.1540  out: 100.00  pv: 223.000  tan: 111.500 →
Setpoint:230.00, Input:223.00, Output:10.00,
 sec: 145.1529  out: 100.00  pv: 0.000  tan: -111.500 →
Setpoint:210.00, Input:0.00, Output:10.00,
 sec: 150.1546  out: 100.00  pv: 223.000  tan: 111.500 →
Setpoint:210.00, Input:212.00, Output:nan,
 sec: 150.1611  out: 100.00  pv: 0.000  tan: -111.500 →

 Controller Action: directIP

 Output Start:      0.00
 Output Step:       100.00
 Sample Sec:        5.0000

 Pv Start:          223.000
 Pv Max:            663.026
 Pv Diff:           440.026

 Process Gain:      25.144
 Dead Time Sec:     0.000
 Tau Sec:           188.705

 Tau/Dead Time:     99.9 (easy to control)
 Tau/Sample Period: 37.7 (good sample rate)

 Tuning Method: ZN_PID
  Kp: inf
  Ki: inf  Ti: nan
  Kd: inf  Td: nan

Updating with new tunings: Kp: inf, Ki: inf, Kd: inf
Setpoint:200.00, Input:191.00, Output:0.00,
Setpoint:225.00, Input:0.00, Output:0.00,
 sec: 155.1621  out: 100.00  pv: 0.000  tan: -111.500 →
 sec: 160.1693  out: 100.00  pv: 0.000  tan: 92.917 ↗
Setpoint:210.00, Input:0.00, Output:10.00,
Setpoint:230.00, Input:223.00, Output:nan,
Setpoint:210.00, Input:212.00, Output:nan,
 sec: 165.1763  out: 100.00  pv: 0.000  tan: -130.083 ↘
Setpoint:200.00, Input:191.00, Output:0.00,
Setpoint:225.00, Input:0.00, Output:0.00,
 sec: 170.1812  out: 100.00  pv: 0.000  tan: 74.333 ↗
 sec: 175.1822  out: 100.00  pv: 0.000  tan: -148.667 ↘
Setpoint:210.00, Input:0.00, Output:10.00,
Setpoint:230.00, Input:223.00, Output:nan,
Setpoint:210.00, Input:212.00, Output:nan,
 sec: 180.1889  out: 100.00  pv: 0.000  tan: 55.750 ↗
Setpoint:200.00, Input:191.00, Output:0.00,
Setpoint:225.00, Input:0.00, Output:0.00,
```

**Comment by FelixWT** [comment]
> In the example you created at wokwi, the led is just blinking and does not stay on.
> 
> Shouldnt it stay on to let the heating heat?

It's blinking because of these two lines

```
const float outputSpan = 1000;
float outputStep = 50;
```

in words this means in 1000 millisec control window, the system will send HIGH signal to pin 22 for 50 millisec.


you might ask how about set this to 

```
const float outputSpan = 1000;
float outputStep = 1000;
```

as far as I know this might lead to the searching of the inflection point to fail , you will reach temperture limit before the search is done

**Comment by FelixWT** [comment]
> Do you have any idea why is my input sometimes 0 when plotting to Serial? I have consistent readings

I'll say this might be the problem on the data collecting end of you system  but I might be wrong
Can you share your code where you assign value to `currentTemp`

**Comment by zekageri** [comment]
**Thank you for your response, i appreciate it.**

the `currentTemp` is coming from a `modbus RTU slave`. It is a thermostat which measures the temperature and i just assign a buffer value from the given index to the `currentTemp` variable. It does not change if i print out my readings directly from the buffer

`currentTemp = modules[thermModule.index]->readBuffer[1];`

the output must stay HIGH since this is the controll bit of the heating system. If this bit is high, the system will start the heating process. It will start to open the valves, starts the boiler and starts the circulation pump after the valves are opened. This is a relatively long process so the controll (output) bit can not fluctuate.

Right now the PID is either left the output bit on high in all cases or does not at all.

**Comment by FelixWT** [comment]
You can try get rid of the PID Library first , and print out temperature every 2 sec. Just for sure that the reading logic of your system is on point.

**Comment by zekageri** [comment]
I have a webserver whichs shows the measured temperature. It is refreshed every 1 sec. I see that it is correct

**Comment by FelixWT** [comment]
I guess I'll provide my code since I'm using ESP32 as well...
note that I'm not using pointers to construct class instance & I'm using relay controled heating system

**main.cpp**
```
#include <Arduino.h>
#include "ArduinoJson.h"
#include "AutoTuneEvent.h"

void setup(){
  Serial.begin(115200);
  Serial2.begin(921600);
  Serial2.setTimeout(10);
  AutoTune.begin();
}

void loop(){
  server.handleClient();
 
  if(Serial2.available()){
    String RCV = Serial2.readString();
    Display_Event.rxDataSet(RCV);
//get data from Json
    StaticJsonDocument<512> doc;
    DeserializationError error = deserializeJson(doc, RCV);
    if (error) {
      String Emsg="DeserializeJson Failed";
      EspNow_Event.send(Emsg);
      return;
    }else{
// Where I assign temperature value to Input   
      float val = doc["val"];
      AutoTune.Input = val;
    }
  }
  AutoTune.Autotuning();
}
```

**AutoTuneEvent.cpp**
```
#include "AutoTuneEvent.h"
#include "LittleFSEvent.h"
#include "ArduinoJson.h"
#include "timerEvent.h"
#include <QuickPID.h>
#include <sTune.h>
#include <Wire.h>

AutoTuneEvent AutoTune = AutoTuneEvent();
sTune tuner = sTune(&AutoTune.Input, &AutoTune.Output, tuner.ZN_PID, tuner.directIP, tuner.printDEBUG);
QuickPID myPID(&AutoTune.Input, &AutoTune.Output, &AutoTune.Setpoint);

//Get Accurate InputSpan
bool AutoTuneEvent::settleInputSpan()
{
  if(Serial2.available()){
    String UART2 = Serial2.readString();
    StaticJsonDocument<512> JBuffer;
    DeserializationError error = deserializeJson(JBuffer, UART2);
    if (error) {
      return false;
    }else{
      float  temp= JBuffer["val"];
      Input = temp;
      inputSpan=Setpoint-temp;
      return true;
    }
  }
  return false;
}

void AutoTuneEvent::begin()
{
  AutoTuneEnable = false;
  NormalTemp = false;
  while(!settleInputSpan());
  Output = 0;
//Heater CMD 
  Serial2.write("{\"HOT\":[0]}");
  delay(3000);
  tuner.Configure(inputSpan, outputSpan, outputStart, outputStep, testTimeSec, settleTimeSec, samples);
  tuner.SetEmergencyStop(tempLimit);
  
  String PID = "";
//Check if the PID Value setting exist in Flash storage as file
  if(readFile(LittleFS,"/PID.txt",PID)){
    PrePIDinFlash = true;
    StaticJsonDocument<128> doc;
    DeserializationError error = deserializeJson(doc, PID);
    if(error){
      PIDExist = false;
    }else{
//Modified Function, Force enter runPID stage
      tuner.setTunerStatus(tuner.TunerStatus::runPid);

      JsonArray PIDJson = doc["PID"];
      Kp = PIDJson[0];
      Ki = PIDJson[1];
      Kd = PIDJson[2];

      myPID.SetOutputLimits(0, outputSpan * 0.5);
      myPID.SetSampleTimeUs((outputSpan-1) * 1000);
      myPID.SetMode(myPID.Control::automatic); // the PID is turned on
      myPID.SetProportionalMode(myPID.pMode::pOnError);
      myPID.SetAntiWindupMode(myPID.iAwMode::iAwOff);
      myPID.SetTunings(Kp, Ki, Kd); // set PID gains
 //Send PID Value Via ESPNOW, send to another esp32 which monitor the Out & Input to generate chart in web page
      char msgBuffer[128];
      int msgLen = snprintf(msgBuffer, sizeof(msgBuffer),"{\"PID\":[%f,%f,%f]}",Kp,Ki,Kd);
      String msg = String(msgBuffer,msgLen);
      EspNow_Event.send(msg);
      PIDExist = true;
    }
  }else{
    PrePIDinFlash = false;
    PIDExist = false;
  }
}

void AutoTuneEvent::Autotuning()
{
  if(!PIDExist){
    tuner.softPwm(Input, Output, Setpoint, outputSpan, 1);
    switch (tuner.Run()) {
      case tuner.sample: {// Finding Inflection Point Stage, Output Stays the same
        String DbMsg = String(tuner.AutoTuneMSG,tuner.MSGLen);
        EspNow_Event.send(DbMsg);
        break;
      }
      case tuner.tunings:{// active just once when sTune is done

        //Assign PID Value According to Selected PID Method
        tuner.SetTuningMethod(tuner.TuningMethod::ZN_PID);
        Kp=tuner.GetKp();
        Ki=tuner.GetKi();
        Kd=tuner.GetKd();

        char msgBuffer[128];
        int msgLen = snprintf(msgBuffer, sizeof(msgBuffer),"{\"PID\":[%f,%f,%f]}",Kp,Ki,Kd);
        String msg = String(msgBuffer,msgLen);
        EspNow_Event.send(msg);

        writeFile(LittleFS,"/PID.txt",msg);
        PIDExist = true;
        //Turn Off Output to Cool Down
        Serial2.write("{\"HOT\":[0]}");
        break;
      }
    }
  }else{
    if(NormalTemp || Input < 40){
      if(!NormalTemp){
        NormalTemp = true;
// Work around logic for Weird output
      }else if(!PrePIDinFlash){
        esp_restart();
      }

      float optimumOutput = tuner.softPwm(Input, Output, Setpoint, outputSpan, 0);
      if(tuner.Run()==tuner.runPid){
        myPID.Compute();
        tuner.plotter(Input, optimumOutput, Setpoint, 0.5f, 3);
        if(Timer_Event.getTmr0_1ms(2000)){
          String DbMsg = String(tuner.AutoTuneMSG,tuner.MSGLen);
          EspNow_Event.send(DbMsg);
        }
      }
    }else{
//wait heater to cool down
      if(Timer_Event.getTmr0_1ms(10000)){
        String DbMsg = String(Input,3);
        EspNow_Event.send(DbMsg);
        char msgBuffer[128];
        int msgLen = snprintf(msgBuffer, sizeof(msgBuffer),"{\"PID\":[%f,%f,%f]}",Kp,Ki,Kd);
        String msg = String(msgBuffer,msgLen);
        EspNow_Event.send(msg);
      }
    }
  }
}
```
**AutoTuneEvent.h**
```
#ifndef AutoTuneEvent_H
#define AutoTuneEvent_H

#include <Arduino.h>

class AutoTuneEvent{
  public:
    /// @brief Prefix Time(sec)
    uint32_t settleTimeSec = 5;
    /// @brief Test Span(sec)
    uint32_t testTimeSec = 5000;
    /// @brief Amount of Samples within Test Span
    uint16_t samples = 10000;
    /// @brief Δ °C from Room Temperature to Temperature Limit
    float inputSpan;
    /// @brief Full Control Span
    float outputSpan = 2000;
    /// @brief Heat Up Span Within Prefix Time
    float outputStart = 0;
    /** Heat Up Span Within Control Span
     * outputStart = 652B ? 500 :( Exp.env ? 100)*/
    float outputStep = 100;
    /** Over Pass This Temperature ? Force Stop
     * tempLimit = 652B ? 320 :( Exp.env ? 210)*/
    float tempLimit = 210;
    bool startup = true;

    /// @brief Use History PID or Not
    bool PIDExist;
    /// @brief Current Temperature Back to Room Temperature Flag.  For Run Time Initial Status Check 
    bool NormalTemp;
    /// @brief Check If the PID Value Just Generated
    bool PrePIDinFlash;

    float Input;
    float Output;
    float Setpoint = 180;

    float Kp;
    float Ki;
    float Kd;

    /// @brief Temperature Sensor Initialize, Tuning Params Set
    void begin();
    /// @brief Run PID Auto Tuning
    void Autotuning();
    /// @brief Calculate The Diff Between Target Temperature & Current Temperature
    bool settleInputSpan();
};

extern AutoTuneEvent AutoTune;
#endif
```
**Modified Function**
```
float sTune::softPwm(float input, float output, float setpoint, uint32_t windowSize, uint8_t debounce) {
  uint32_t msNow = millis();
  static uint32_t  windowStartTime, nextSwitchTime;
  if (msNow - windowStartTime >= windowSize) {
    windowStartTime = msNow;
  }
  static float optimumOutput;
  static bool reachedSetpoint;

  if (input > setpoint){
    reachedSetpoint = true;
  }

  optimumOutput = output<0 ? 0:output;

  // Output
  static bool relayStatus;
  if (!relayStatus && optimumOutput > (msNow - windowStartTime)) {
    if (msNow > nextSwitchTime) {
      nextSwitchTime = msNow + debounce;
      relayStatus = true;
      snprintf(HeatUpSpan, sizeof(HeatUpSpan),"{\"HOT\":[%.3f]}",optimumOutput);
      Serial.println(HeatUpSpan);
      Serial2.write(HeatUpSpan);
    }
  } else if (relayStatus && optimumOutput < (msNow - windowStartTime)) {
    if (msNow > nextSwitchTime) {
      nextSwitchTime = msNow + debounce;
      relayStatus = false;
    }
  }
  return optimumOutput;
}
```
**Debug Message**
```
#1528 Sec: 764.0610 Out: 100 PV: 103.200 ipCount: 610 tan: 14.156  ↘
#1529 Sec: 764.5610 Out: 100 PV: 103.300 ipCount: 611 tan: 14.201  ↗
#1530 Sec: 765.0610 Out: 100 PV: 103.300 ipCount: 612 tan: 14.246  ↗
#1531 Sec: 765.5610 Out: 100 PV: 103.400 ipCount: 613 tan: 14.191  ↘
#1532 Sec: 766.0610 Out: 100 PV: 103.500 ipCount: 614 tan: 14.137  ↘
#1533 Sec: 766.5610 Out: 100 PV: 103.500 ipCount: 615 tan: 14.082  ↘
#1534 Sec: 767.0610 Out: 100 PV: 103.500 ipCount: 616 tan: 14.127  ↗
#1535 Sec: 767.5610 Out: 100 PV: 103.500 ipCount: 617 tan: 14.172  ↗
#1536 Sec: 768.0610 Out: 100 PV: 103.500 ipCount: 618 tan: 14.117  ↘
#1537 Sec: 768.5610 Out: 100 PV: 103.600 ipCount: 619 tan: 14.162  ↗
#1538 Sec: 769.0610 Out: 100 PV: 103.600 ipCount: 620 tan: 14.207  ↗
#1539 Sec: 769.5610 Out: 100 PV: 103.600 ipCount: 621 tan: 14.152  ↘
#1540 Sec: 770.0610 Out: 100 PV: 103.600 ipCount: 622 tan: 14.197  ↗
#1541 Sec: 770.5610 Out: 100 PV: 103.700 ipCount: 623 tan: 14.142  ↘
#1542 Sec: 771.0610 Out: 100 PV: 103.700 ipCount: 624 tan: 14.086  ↘
=== Tuning Method: ZN_PID ===
Kp: 1.020781
Ki: 0.006172   Ti: 165.380737
Kd: 0.024689 Td: 41.345184

=== Tuning Method: Damped_PID ===
Kp: 1.122859
Ki: 0.001618   Ti: 693.915344
Kd: 0.004045 Td: 277.566132

=== Tuning Method: NoOvershoot_PID ===
Kp: 1.020781
Ki: 0.000449   Ti: 2270.995605
Kd: 0.024689 Td: 41.345184

=== Tuning Method: CohenCoon_PID ===
Kp: 21.616938
Ki: 0.003989   Ti: 5419.791016
Kd: 0.034172 Td: 632.584229

=== Tuning Method: Mixed_PID ===
Kp: 6.195339
Ki: 0.003057   Ti: 2026.531128
Kd: 0.021899 Td: 282.904297

=== Tuning Method: ZN_PI ===
Kp: 0.765585
Ki: 0.003703   Ti: 206.723846
Kd: 0.000000 Td: 0.000000

=== Tuning Method: Damped_PI ===
Kp: 0.842144
Ki: 0.001169   Ti: 720.604309
Kd: 0.000000 Td: 0.000000

=== Tuning Method: NoOvershoot_PI ===
Kp: 0.595455
Ki: 0.000375   Ti: 1589.696899
Kd: 0.000000 Td: 0.000000

=== Tuning Method: CohenCoon_PI ===
Kp: 14.577548
Ki: 0.003989   Ti: 3654.877441
Kd: 0.000000 Td: 0.000000

=== Tuning Method: Mixed_PI ===
Kp: 4.195183
Ki: 0.002309   Ti: 1817.045654
Kd: 0.000000 Td: 0.000000

103.800
{"PID":[1.020781,0.006172,0.024689]}
103.900
{"PID":[1.020781,0.006172,0.024689]}
104.100
{"PID":[1.020781,0.006172,0.024689]}
103.800
{"PID":[1.020781,0.006172,0.024689]}
102.900
{"PID":[1.020781,0.006172,0.024689]}
101.600
{"PID":[1.020781,0.006172,0.024689]}
```
Hope this help

**Comment by zekageri** [comment]
Thank you for the snippet. I will read it carefully! :D

**Comment by FelixWT** [comment]
I'll recommand you don't look it to following part, the logic wasn't  establish correctly

**AutoTuneEvent::begin()**

> myPID.SetMode(myPID.Control::automatic); // the PID is turned on
> myPID.SetProportionalMode(myPID.pMode::pOnError);
> myPID.SetAntiWindupMode(myPID.iAwMode::iAwOff);

**AutoTuneEvent::Autotuning()**
> else{
>     if(NormalTemp || Input < 40){
>       if(!NormalTemp){
>         NormalTemp = true;
> // Work around logic for Weird output
>       }else if(!PrePIDinFlash){
>         esp_restart();
>       }
> 
>       float optimumOutput = tuner.softPwm(Input, Output, Setpoint, outputSpan, 0);
>       if(tuner.Run()==tuner.runPid){
>         myPID.Compute();
>         tuner.plotter(Input, optimumOutput, Setpoint, 0.5f, 3);
>         if(Timer_Event.getTmr0_1ms(2000)){
>           String DbMsg = String(tuner.AutoTuneMSG,tuner.MSGLen);
>           EspNow_Event.send(DbMsg);
>         }
>       }
>     }else{
> //wait heater to cool down
>       if(Timer_Event.getTmr0_1ms(10000)){
>         String DbMsg = String(Input,3);
>         EspNow_Event.send(DbMsg);
>         char msgBuffer[128];
>         int msgLen = snprintf(msgBuffer, sizeof(msgBuffer),"{\"PID\":[%f,%f,%f]}",Kp,Ki,Kd);
>         String msg = String(msgBuffer,msgLen);
>         EspNow_Event.send(msg);
>       }
>     }
>   }

**Comment by zekageri** [comment]
Thank you for the clarification. I read trought your sketch, however i can not adapt it to my system properly.
No matter what option i update, the PID always doing a PWM on my controll variable.

I adapted my `pidLoop()` this way:

```cpp
void thermSystem::adaptInputSpan(){
    if(prevSetPoint != Setpoint){
        prevSetPoint = Setpoint;
        inputSpan = Setpoint - Input;
        tuner->Configure(
            inputSpan,
            outputSpan,
            outputStart,
            outputStep,
            testTimeSec,
            settleTimeSec,
            samples
        );
    }
}

void thermSystem::pidLoop() {
    if(!isPidEnabled || !pidInited ){return;}
    adaptInputSpan();
    tuner->softDigit(shouldStartHeatCtr, Input, Output, Setpoint, outputSpan, 1);

    if( shouldStartHeat && !shouldStartHeatCtr){
        if( millis() - lastShouldStartHeatCtr >= 50000 ){
            shouldStartHeat = false;
            lastShouldStartHeatCtr = millis();
        }
    }else if( !shouldStartHeat && shouldStartHeatCtr ){
        if( millis() - lastShouldStartHeatCtr >= 50000 ){
            shouldStartHeat = true;
            lastShouldStartHeatCtr = millis();
        }
    }

    switch (tuner->Run()) {
        case tuner->sample:  // active once per sample during test
            Input = currentTemp;
            tuner->plotter(Input, Output * 0.1, Setpoint, 1, 3);
            break;

        case tuner->tunings:    // active just once when sTune is done
            tuner->GetAutoTunings(&Kp, &Ki, &Kd);
            myPID->SetTunings(Kp, Ki, Kd);
            Serial.printf("Updating with new tunings: Kp: %0.2f, Ki: %0.2f, Kd: %0.2f\n",Kp,Ki,Kd);
            break;

        case tuner->runPid: // active once per sample after tunings
            Input = currentTemp;
            myPID->Compute();
            tuner->plotter(Input, Output, Setpoint, 0.1f, 3);
            break;
    }
}
```

`adaptInputSpan()` will check if the set temperature has changed, and if it is, it will re configure the tuner with the new inputSpan. This however did not help on the digital pwm. I added these lines:

```cpp
if( shouldStartHeat && !shouldStartHeatCtr){
    if( millis() - lastShouldStartHeatCtr >= 50000 ){
        shouldStartHeat = false;
        lastShouldStartHeatCtr = millis();
    }
}else if( !shouldStartHeat && shouldStartHeatCtr ){
    if( millis() - lastShouldStartHeatCtr >= 50000 ){
        shouldStartHeat = true;
        lastShouldStartHeatCtr = millis();
    }
}
```

to prevent the `shouldStartHeat` variable from constant change. It did helped a bit but the pid is not reliable.

Here is my outputs:

```cpp
Setpoint:195.00, Input:173.00, Output:85.00,
 sec: 188.4460  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 188.9470  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 189.4470  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
Setpoint:195.00, Input:173.00, Output:85.00,
 sec: 189.9488  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 190.4488  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 190.9490  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
Setpoint:195.00, Input:173.00, Output:85.00,
 sec: 191.4500  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 191.9520  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 192.4550  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
Setpoint:195.00, Input:173.00, Output:85.00,
 sec: 192.9570  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 193.4578  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 193.9590  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
Setpoint:195.00, Input:173.00, Output:85.00,
 sec: 194.4610  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 194.9610  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 195.4618  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
Setpoint:195.00, Input:173.00, Output:85.00,
 sec: 195.9620  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 196.4630  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 196.9633  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
Setpoint:195.00, Input:173.00, Output:85.00,
 sec: 197.4640  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 197.9650  out: 850.00  pv: 173.000  ipCount: 1  tan: 0.000 ⚠ →
 sec: 198.4660  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.002 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 198.9670  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.003 ⚠ ↗
 sec: 199.4670  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.005 ⚠ ↗
 sec: 199.9670  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.007 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 200.4690  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.008 ⚠ ↗

[MAIN] - OPENING VALVES.
[MAIN] - OPENING VALVES

Opening valve: szelep with id: i_lechzzb1_ypf2n
Opening valve: szelep 2 with id: b_lecjm6hq_vw23b

 sec: 200.9700  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.010 ⚠ ↗
 sec: 201.4709  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.012 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 201.9710  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.013 ⚠ ↗
 sec: 202.4720  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.015 ⚠ ↗
 sec: 202.9740  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.017 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 203.4759  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.018 ⚠ ↗
 sec: 203.9770  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.020 ⚠ ↗
 sec: 204.4790  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.022 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 204.9798  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.023 ⚠ ↗
 sec: 205.4800  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.025 ⚠ ↗

[MAIN] - VALVES OPENED. STARTING BOILER AND PUMPS
[MAIN] - STARTING PUMPS

 sec: 205.9809  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.027 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 206.4830  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.028 ⚠ ↗
 sec: 206.9840  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.030 ⚠ ↗
 sec: 207.4850  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.032 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 207.9850  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.033 ⚠ ↗
 sec: 208.4850  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.035 ⚠ ↗
 sec: 208.9850  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.037 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 209.4868  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.038 ⚠ ↗
 sec: 209.9875  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.040 ⚠ ↗
 sec: 210.4890  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.042 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 210.9910  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.043 ⚠ ↗
 sec: 211.4910  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.045 ⚠ ↗
 sec: 211.9920  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.047 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 212.4928  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.048 ⚠ ↗
 sec: 212.9930  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.050 ⚠ ↗
 sec: 213.4930  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.052 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 213.9930  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.053 ⚠ ↗
 sec: 214.4950  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.055 ⚠ ↗
 sec: 214.9960  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.057 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 215.4978  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.058 ⚠ ↗
 sec: 215.9990  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.060 ⚠ ↗
 sec: 216.5000  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.062 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 217.0010  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.063 ⚠ ↗
 sec: 217.5030  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.065 ⚠ ↗
 sec: 218.0040  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.067 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 218.5040  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.068 ⚠ ↗
 sec: 219.0049  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.070 ⚠ ↗
 sec: 219.5060  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.072 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 220.0070  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.073 ⚠ ↗
 sec: 220.5070  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.075 ⚠ ↗
 sec: 221.0090  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.077 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 221.5100  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.078 ⚠ ↗
 sec: 222.0110  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.080 ⚠ ↗
 sec: 222.5120  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.082 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 223.0120  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.083 ⚠ ↗
 sec: 223.5120  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.085 ⚠ ↗
 sec: 224.0140  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.087 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 224.5140  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.088 ⚠ ↗
 sec: 225.0160  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.090 ⚠ ↗
 sec: 225.5170  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.092 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 226.0190  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.093 ⚠ ↗
 sec: 226.5208  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.095 ⚠ ↗
 sec: 227.0220  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.097 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
 sec: 227.5220  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.098 ⚠ ↗
 sec: 228.0230  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.100 ⚠ ↗
 sec: 228.5240  out: 850.00  pv: 174.000  ipCount: 1  tan: 0.102 ⚠ ↗
Setpoint:195.00, Input:174.00, Output:85.00,
```

It just can't learn the valve open time.
Tomorrow i will build a small test environment with a temperature sensor where my system can controll the heat inside.
It will be a big box with an old wolfram light bulb which can produce heat so it will heat up the box. I hope i can test reliably with that.

Here are my options for the pid

```cpp
void adaptInputSpan();
uint32_t settleTimeSec  = 5;
uint32_t testTimeSec    = 5000;
uint16_t samples    = 10000;
float inputSpan     = 350;
float outputSpan    = 1000;
float outputStart   = 0;
float outputStep    = 850;
float tempLimit     = 350;
float prevSetPoint  = 0;

// variables
float Input, Output, Setpoint = 50, Kp, Ki, Kd;
sTune *tuner;
QuickPID *myPID;

void pidInit();
void pidLoop();
/* PID DETAILS */

void checkHeatingStatus();
boolean shouldStartHeat = false;
boolean shouldStartHeatCtr = false;
long lastShouldStartHeatCtr = 0;
```

**Comment by zekageri** [comment]
Here is a wokwi which tries to demonstrate my environmnet.

https://wokwi.com/projects/357187661998047233

You can clearly see that if it hits the set temperature it will heat even further and did not stop the process.
And the pid will blow off the heating as soon as it can.

**Comment by FelixWT** [comment]
You closed this issue, so I'm assuming you've got it working now

My apologies, I may have misled you with the wrong code. The definition of inputSpan should be: 
```
/// @brief Δ °C from Room Temperature to Temperature Limit
float inputSpan;
```
if you look in to to library you can find following equation:
https://github.com/Dlloydev/sTune/blob/94147e1a91036672667f186089614b0db9c525ed/src/sTune.cpp#L194
which is the calculation of process gain, thus inputSpan should be defined as mentioned

---------------------------------------------------------
Little update of my fuction:
```
/// @brief Timer Interrupt Controlled Output
/// @param pidAva PID Available
void sTune::softPwm(bool pidAva, float setpoint, uint32_t windowSize) {
  if(Timer_Event.getTmr0_1ms(windowSize)){
    if(!pidAva){
      snprintf(HeatUpSpan, sizeof(HeatUpSpan),"{\"HOT\":[%.3f]}",*_output);
      Serial.println(HeatUpSpan);
      //Send Relay NC timespan via UART
      Serial2.write(HeatUpSpan);
    }
    OutSpanReach=true;
  }
}
```

**Comment by zekageri** [comment]
Thank you for the response. It does not work, i closed this issue to open a new one with my clarifications.
Here is the new issue #29 
PID does not work at all in my environmnet right now.
